NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.cpp

```diff
@@ -0,0 +1,173 @@
+}
+
+void LocalLinearLeastSquaresExtrapolator::calculateResiduals(
+    ExtrapolatableElementCollection const& extrapolatables)
+{
+    assert(static_cast<std::size_t>(_residuals.size()) ==
+           extrapolatables.size());
+
+    auto const size = extrapolatables.size();
+    for (std::size_t i=0; i<size; ++i) {
+        calculateResiudalElement(i, extrapolatables);
+    }
+}
+
+void LocalLinearLeastSquaresExtrapolator::extrapolateElement(
+    std::size_t const element_index,
+    ExtrapolatableElementCollection const& extrapolatables,
+    GlobalVector& counts)
+{
+    auto const& integration_point_values =
+        extrapolatables.getIntegrationPointValues(
+            element_index, _integration_point_values_cache);
+
+    // number of nodes in the element
+    const auto nn = extrapolatables.getShapeMatrix(element_index, 0).cols();
+    // number of integration points in the element
+    const auto ni = integration_point_values.size();
+
+    assert(ni >= static_cast<decltype(ni)>(nn) &&
+           "Least squares is not possible if there are more nodes than"
+           "integration points.");
+
+    auto& N = _local_matrix_cache; // TODO make that local?
+    N.resize(ni, nn); // TODO: might reallocate very often
+
+    for (auto int_pt = decltype(ni){0}; int_pt < ni; ++int_pt) {
+        auto const& shp_mat =
+            extrapolatables.getShapeMatrix(element_index, int_pt);
+        assert(shp_mat.cols() == nn);
+
+        // copy shape matrix to extrapolation matrix columnwise
+        N.block(int_pt, 0, 1, nn) = shp_mat;
+    }
+
+    // TODO make gp_vals an Eigen::VectorXd const& ?
+    Eigen::Map<const Eigen::VectorXd> const integration_point_values_vec(
+        integration_point_values.data(), integration_point_values.size());
+
+    // TODO
+    // optimization: Store decomposition of N*N^T or N^T for reuse?
+    //   cf. http://eigen.tuxfamily.org/dox/classEigen_1_1FullPivLU.html
+    //   cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html
+    //   cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
+    //   cf. https://eigen.tuxfamily.org/dox/classEigen_1_1HouseholderQR.html
+    // Extrapolate several values at once?
+
+    // do the least squares computation using QR decomposition.
+    Eigen::VectorXd tmp = N.householderQr().solve(integration_point_values_vec);
+
+    // option: do the least squares computation using LDLT decomposition.
+    // Eigen::VectorXd tmp =
+    //         (N*N.transpose()).ldlt().solve(N*integration_point_values_vec);
+
+    // TODO: for now always zeroth component is used
+    auto const& global_indices = _local_to_global(element_index, 0).rows;
+
+    _nodal_values.add(global_indices,
+                      tmp);  // TODO does that give rise to PETSc problems?
+    counts.add(global_indices, std::vector<double>(global_indices.size(), 1.0));
+}
+
+void LocalLinearLeastSquaresExtrapolator::calculateResiudalElement(
```

Comment:
- typo: residual
 - Could you give a short comment on how (mathematically) the residual is computed?